### DISCUSSION TITLE
@ReadOnly doesnt seem to apply on @VirtualRef property. Or am I missing something?

### DISCUSSION THREAD

**Discussion by alexdonh** [TypedProject/ts-express-decorators#1605] [Q&A] [answered]
Hi @Romakita 

Is it only me or the `@ReadOnly` decorator doesnt seem to apply on propreties that have `@VirtualRef` on? Maybe I am just missing something? Or, is there a way to make a virtual ref property readonly?  (Not appear on request body - but on response)

**Comment by Romakita** [comment, marked as answer]
Hello @alexdonh 
Readonly is just for swagger spec. The metadata isn’t consumed by mongoose or the endpoint. The official decorator to use a model with multiple output models is `@Groups`.

see the documentation on this decorator please. See you
Romain

**Reply by alexdonh** [reply]
It isn't about multiple output models. It's about a model used for both input and output. Imagine this model:

```typescript
class User {
  @ObjectID("id")
  @ReadOnly()
  _id: ObjectID;

  @Format("date-time")
  @Default(Date.now())
  @ReadOnly()
  createdAt: number = Date.now();

  @Format("date-time")
  @Default(Date.now())
  @ReadOnly()
  updatedAt: number = Date.now();

  @Required()
  name: string;

  @Required()
  @Email()
  email: string;

  @Ignore()
  groupId: ObjectID;

  @ReadOnly()
  @VirtualRef({
    ref: Group,
    localField: "groupId",
    foreignField: "_id",
    justOne: true
  })
  group: VirtualRef<Group>;
}
```

In swagger request body it would look like

```typescript
{
  "name": "string",
  "email": "string",
  "group": {
     ...
  }
}
```

and in response

```typescript
{
  "id": "5ce7ad3028890bd71749d477",
  "createdAt": 1635301703658,
  "updatedAt": 1635301703658,
  "name": "string",
  "email": "string",
  "group": {
     ...
  }
}
```

(In controller I use `User` class in both `@BodyParams` and `@Returns`) 

As you can see, `@ReadOnly` works as expected for `id`, `createdAt` and `updatedAt` but `group`. If I remove `@VirtualRef` decorator then it works as expected.

I can use `@Groups` to handle it like this

```typescript
  @Groups("!add")
  @VirtualRef({
    ref: Group,
    localField: "groupId",
    foreignField: "_id",
    justOne: true
  })
  group: VirtualRef<Group>;
```

then in Controller `@BodyParams` 

```typescript
async add(
  @Required()
  @BodyParams()
  @Groups("add")
  params: User
) {
 ...
}
```

Is it expected to work this way? Does it mean every time we want to "hide" properties with `@VirtualRef` we need `@Groups`?

**Reply by Romakita** [reply]
Hoo ok. you just taught me how readOnly works in swagger :)

So you totally right, the Readonly should do the job for VirtualRef also. But VirtualRef decorator seems to be special:
https://github.com/tsedio/tsed/blob/production/packages/orm/mongoose/src/decorators/virtualRef.ts 
It thinks the decorator order will be important.

One question: VirtualRef must be always ReadOnly ?

**Reply by Romakita** [reply]
I found that: https://stackoverflow.com/questions/51402156/how-to-declare-a-ref-property-as-readonly-in-openapi-swagger

$ref used with readOnly is only possible with OpenAPI 3.1 without workaround (but this version is not released).

I created a PR to fix that: https://github.com/tsedio/tsed/pull/1607/files

**Reply by Romakita** [reply]
Released ;)

**Reply by alexdonh** [reply]
Been off for a few days. I am so glad that you ve always been dedicating to TsED. Thanks for the quick response. Appreciated it!
